daemon/emer-daemon.c

```diff
@@ -1123,6 +1145,169 @@ set_max_bytes_buffered (EmerDaemon *self,
   priv->max_bytes_buffered = max_bytes_buffered;
 }
 
+static void schedule_next_midnight_tick (EmerDaemon *self);
+static void schedule_next_month_tick (EmerDaemon *self);
+
+static void
+save_aggregate_timers_to_tally (EmerDaemon *self,
+                                const char *date)
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  EmerAggregateTimerImpl *timer_impl;
+  GHashTableIter iter;
+  gint64 now_monotonic_us;
+
+  now_monotonic_us = g_get_monotonic_time ();
+
+  g_hash_table_iter_init (&iter, priv->aggregate_timers);
+  while (g_hash_table_iter_next (&iter, (gpointer *) &timer_impl, NULL))
+    {
+      emer_aggregate_timer_impl_store (timer_impl,
+                                       date,
+                                       now_monotonic_us);
+      emer_aggregate_timer_impl_split (timer_impl,
+                                       now_monotonic_us);
+    }
+}
+
+static EmerTallyIterResult
+buffer_daily_aggregate_event_to_queue (guint32     unix_user_id,
+                                       GVariant   *event_id,
+                                       GVariant   *aggregate_key,
+                                       GVariant   *payload,
+                                       guint32     counter,
+                                       const char *date,
+                                       gpointer    user_data)
+{
+  EmerDaemon *self = user_data;
+
+  emer_daemon_record_aggregate_event (self,
+                                      event_id,
+                                      date,
+                                      counter,
+                                      payload != NULL,
+                                      payload);
+
+  return EMER_TALLY_ITER_CONTINUE;
+}
+
+static void
+buffer_daily_aggregate_events_to_queue (EmerDaemon *self)
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  g_autofree gchar *date = NULL;
+
+  date = g_date_time_format (priv->current_aggregate_tally_date,
+                             "%Y-%m-%d");
+
+  emer_aggregate_tally_iter (priv->aggregate_tally,
+                             date,
+                             buffer_daily_aggregate_event_to_queue,
+                             self);
+}
+
+static gboolean
+clock_ticked_midnight_cb (gpointer user_data)
+{
+  EmerDaemon *self = user_data;
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  g_autofree char *date = NULL;
+
+  date = g_date_time_format (priv->current_aggregate_tally_date,
+                             "%Y-%m-%d");
+
+  save_aggregate_timers_to_tally (self, date);
+  buffer_daily_aggregate_events_to_queue (self);
+  schedule_next_midnight_tick (self);
+
+  return G_SOURCE_REMOVE;
+}
+
+static gboolean
+clock_turned_month_cb (gpointer user_data)
```

Comment:
I don't think you need two timers. You can update the current/just-ended month's tallies at the same time as the midnight tick